domain/secretbackend/state/operation.go

```diff
@@ -0,0 +1,202 @@
+		}
+		if o.BackendType == "" {
+			return fmt.Errorf("backend type is missing: %w", backenderrors.NotValid)
+		}
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+	// Update.
+	if existing.BackendType == "" {
+		// This should never happen, but just in case.
+		return errors.Errorf("backend type is empty for backend %q", o.ID)
+	}
+	if existing.Name == "" {
+		// This should never happen, but just in case.
+		return errors.Errorf("backend name is empty for backend %q", o.ID)
+	}
+	if o.BackendType != "" && o.BackendType != existing.BackendType {
+		// The secret backend type is immutable.
+		return errors.Errorf(
+			"cannot change backend type from %q to %q because backend type is immutable",
+			existing.BackendType, o.BackendType,
+		)
+	}
+	// Fill in the existing backend type.
+	o.BackendType = existing.BackendType
+	if o.Name == "" {
+		// Fill in the existing name.
+		o.Name = existing.Name
+	}
+	if o.TokenRotateInterval == nil && existing.TokenRotateInterval.Valid {
+		// Fill in the existing token rotate interval.
+		o.TokenRotateInterval = &existing.TokenRotateInterval.Duration
+	}
+	return nil
+}
+
+// Apply applies the upsert operation to the database.
+func (o *upsertOperation) Apply(ctx context.Context, tx *sqlair.TX) error {
+	if err := o.validate(ctx, tx); err != nil {
+		return errors.Trace(err)
+	}
+	sb := SecretBackend{
+		ID:          o.ID,
+		Name:        o.Name,
+		BackendType: o.BackendType,
+	}
+	if o.TokenRotateInterval != nil {
+		sb.TokenRotateInterval = domain.NullableDuration{Duration: *o.TokenRotateInterval, Valid: true}
+	}
+
+	err := tx.Query(ctx, o.upsertBackendStmt, sb).Run()
+	if database.IsErrConstraintUnique(err) {
+		return fmt.Errorf("secret backend name %q: %w", sb.Name, backenderrors.AlreadyExists)
+	}
+	if err != nil {
+		return fmt.Errorf("cannot upsert secret backend %q: %w", o.Name, err)
+	}
+	if o.NextRotateTime != nil && !o.NextRotateTime.IsZero() {
+		err = tx.Query(ctx, o.upsertRotationStmt, SecretBackendRotation{
+			ID:               o.ID,
+			NextRotationTime: sql.NullTime{Time: *o.NextRotateTime, Valid: true},
+		}).Run()
+		if err != nil {
+			return fmt.Errorf("cannot upsert secret backend rotation time for %q: %w", o.Name, err)
+		}
+	}
+	if len(o.Config) == 0 {
+		return nil
+	}
+	namesToKeep := make(sqlair.S, 0, len(o.Config))
+	for k := range o.Config {
+		namesToKeep = append(namesToKeep, k)
+	}
+	err = tx.Query(ctx, o.clearConfigStmt, sqlair.M{"uuid": o.ID}, namesToKeep).Run()
+	if err != nil {
+		return fmt.Errorf("cannot clear secret backend config for %q: %w", o.Name, err)
+	}
+	for k, v := range o.Config {
+		err = tx.Query(ctx, o.upsertConfigStmt, SecretBackendConfig{
+			ID:      o.ID,
+			Name:    k,
+			Content: v.(string),
+		}).Run()
+		if database.IsErrConstraintCheck(err) {
+			return fmt.Errorf(
+				"cannot upsert secret backend config for %q: %w",
+				o.Name, errors.NewNotValid(nil, "empty config name or content"),
```

Comment:
We should use a typed error. But we can avoid this if we check if `v` is empty or not. We'd do that in the validate() so by the time we get to here, we know v is never empty.